extras/upgrade_fw.py

```diff
@@ -0,0 +1,161 @@
+                if dev.fw_update_type in upgrade_type
+                and any(x == vendor_string for x in dev.vendor)
+            ][0]
+            logger.info("%s is a %s %s" % (ip, vendor_string, dev.__name__))
+        except IndexError:
+            err_msg = "Cannot find a proper Device class for %s: %s" % (
+                vendor_string,
+                Dmi.chassis_names[type_index],
+            )
+            logger.error(err_msg)
+            raise RuntimeError(err_msg)
+
+        if issubclass(dev, LVFSDevice):
+            return dev(ip, user, password)
+        elif issubclass(dev, OEMDevice):
+            if not (
+                "bmc_ip" in options
+                and "bmc_user" in options
+                and "bmc_password" in options
+            ):
+                raise RuntimeError(
+                    "Please provide $BMC_IP, $BMC_USER, $BMC_PASSWORD for this device"
+                )
+            exit()
+            return dev(
+                options.get("bmc_ip"),
+                options.get("bmc_user"),
+                options.get("bmc_password"),
+            )
+    except subprocess.CalledProcessError as e:
+        logger.error(e.output)
+        raise RuntimeError(e.output)
+
+
+def main():
+    """
+    mandantory parameter: $DEVICE_IP (testflinger env variable)
+    extra parameters for Server: $BMC_IP, $BMC_USERNAME, $BMC_PASSWORD
+    (not yet provided by testflinger)
+    """
+    parser = argparse.ArgumentParser(
+        epilog="Example: %(prog)s upgrade 10.0.0.1",
+        usage="%(prog)s {upgrade, downgrade, detect} $DEVICE_IP [options]",
+    )
+    parser.add_argument(
+        choices=["upgrade", "downgrade", "detect"],
+        dest="action",
+        help="select to upgrade firmware, downgrade firmware, or just detect DUT info without further actions",
+    )
+    parser.add_argument("device_ip", help="$DEVICE_IP")
+    parser.add_argument(
+        "--bmc_ip",
+        "-i",
+        nargs="?",
+    )
+    parser.add_argument(
+        "--bmc_user",
+        "-u",
+        nargs="?",
+    )
+    parser.add_argument(
+        "--bmc_password",
+        "-p",
+        nargs="?",
+    )
+    args = parser.parse_args()
+
+    log_file = "/tmp/upgrade_fw.log"
+    logger.basicConfig(
+        level=logging.DEBUG,
+        format="%(asctime)s %(levelname)-8s %(message)s",
+        datefmt="%y-%m-%d %H:%M:%S",
+        filename=log_file,
+        filemode="w",
+    )
+
+    if args.bmc_ip and args.bmc_user and args.bmc_password:
+        target_device = detect_device(
+            args.device_ip,
+            target_device_username,
+            options=vars(args),
+        )
+    else:
+        target_device = detect_device(
+            args.device_ip, target_device_username, password="insecure"
+        )
+
+    target_device.get_fw_info()
+    if args.action == "detect":
+        print("Check %s for details." % log_file)
+        return
+    elif args.action == "upgrade":
```

Comment:
No need to use “else if (elif)” after return. You can just use a new if statement :)

Reply:
Fixed.